pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package applicationconfiguration
+
+import (
+	"context"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+
+	"github.com/crossplane/crossplane/apis/oam/v1alpha2"
+)
+
+// Render error strings.
+const (
+	errUnmarshalWorkload = "cannot unmarshal workload"
+	errUnmarshalTrait    = "cannot unmarshal trait"
+)
+
+// Render error format strings.
+const (
+	errFmtGetComponent     = "cannot get component %q"
+	errFmtResolveParams    = "cannot resolve parameter values for component %q"
+	errFmtRenderWorkload   = "cannot render workload for component %q"
+	errFmtRenderTrait      = "cannot render trait for component %q"
+	errFmtSetParam         = "cannot set parameter %q"
+	errFmtUnsupportedParam = "unsupported parameter %q"
+	errFmtRequiredParam    = "required parameter %q not specified"
+)
+
+// A ComponentRenderer renders an ApplicationConfiguration's Components into
+// workloads and traits.
+type ComponentRenderer interface {
+	Render(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error)
+}
+
+// A ComponentRenderFn renders an ApplicationConfiguration's Components into
+// workloads and traits.
+type ComponentRenderFn func(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error)
+
+// Render an ApplicationConfiguration's Components into workloads and traits.
+func (fn ComponentRenderFn) Render(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error) {
+	return fn(ctx, ac)
+}
+
+type components struct {
+	client   client.Reader
+	params   ParameterResolver
+	workload ResourceRenderer
+	trait    ResourceRenderer
```

Comment:
Because `r.parameterResolver.Resolve()` [stutters](https://michaelwhatcott.com/go-code-that-stutters/). I believe the code is more readable as `r.params.Resolve` reads more like "resolve parameters".